src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -32,48 +34,50 @@ const vscodeNotebookEnums = require('vscode') as typeof import('vscode-proposed'
 export class KernelExecution implements IDisposable {
     private readonly cellExecutions = new WeakMap<NotebookCell, CellExecution>();
 
-    private readonly documentExecutions = new WeakMap<NotebookDocument, MultiCancellationTokenSource>();
+    private readonly documentExecutions = new WeakMap<NotebookDocument, CancellationTokenSource>();
+    private readonly stackOfCellsToExecuteByDocument = new WeakMap<NotebookDocument, CellExecution[]>();
 
     private readonly kernelValidated = new WeakMap<NotebookDocument, { kernel: IKernel; promise: Promise<void> }>();
 
     private readonly executionFactory: CellExecutionFactory;
     private readonly disposables: IDisposable[] = [];
+    private readonly pendingExecution = new ChainedExecutions<void>('Kernel execution failure');
+    private readonly pendingCellExecution = new ChainedExecutions<NotebookCellRunState | undefined>(
+        'Kernel execution previous cell failure'
+    );
     private isRawNotebookSupported?: Promise<boolean>;
+    private _interruptPromise?: Promise<InterruptResult>;
     constructor(
         private readonly kernelProvider: IKernelProvider,
-        private readonly commandManager: ICommandManager,
         errorHandler: IDataScienceErrorHandler,
         editorProvider: INotebookEditorProvider,
         readonly kernelSelectionUsage: IKernelSelectionUsage,
         readonly appShell: IApplicationShell,
         readonly vscNotebook: IVSCodeNotebook,
         readonly metadata: Readonly<KernelConnectionMetadata>,
         private readonly rawNotebookSupported: IRawNotebookSupportedService,
-        context: IExtensionContext
+        context: IExtensionContext,
+        private readonly interruptTimeout: number
     ) {
         this.executionFactory = new CellExecutionFactory(errorHandler, editorProvider, appShell, vscNotebook, context);
     }
 
     @captureTelemetry(Telemetry.ExecuteNativeCell, undefined, true)
     public async executeCell(notebookPromise: Promise<INotebook>, cell: NotebookCell): Promise<void> {
-        // Cannot execute empty cells.
-        if (this.cellExecutions.has(cell) || cell.document.getText().trim().length === 0) {
-            // clear cell output
-            cell.outputs = [];
```

Comment:
This was incorrect, we must use the VSC Edit API to update the cell output.
Moved updating this into cellExecution. So that everything goes through cell execution (consistent way to update output).